File: Libraries/rust/beagle_helper/src/devices/light_sensor.rs
use crate::{abstractions::adc::Adc, pin::Pin};

pub struct LightSensor {
    adc: Adc,
    threshold: f64,
}

impl LightSensor {
    pub fn new(pin: Pin, threshold: Option<f64>) -> std::io::Result<Self> {
        let adc = pin.adc_input()?;
        Ok(Self {
            adc,
            threshold: threshold.unwrap_or(0.1),
        })
    }

    pub fn value(&mut self) -> std::io::Result<bool> {
        let ref_voltage = self.adc.ref_voltage();
        let scaled = self.adc.read_scaled()?;

        Ok((scaled / ref_voltage) > self.threshold)
    }

    pub fn wait_for_dark(&mut self) {
        loop {
            if let Ok(false) = self.value() {
                return;
            }
        }
    }

    pub fn wait_for_light(&mut self) {
        loop {
            if let Ok(true) = self.value() {
                return;
            }
        }
    }
}

File: Libraries/python/beagle_helper/src/beagle_helper/devices/light_sensor.py
from ..pin import Pin

class LightSensor:
    def __init__(self, pin: Pin, threashold: float | None = None) -> None:
        self.adc = pin.adc_input()
        if threashold:
            self.threashold = threashold
        else:
            self.threashold = 0.1


    def value(self) -> bool:
        ref_v = self.adc.ref_voltage()
        scaled = self.adc.read_scaled()

        return (scaled / ref_v) > self.threashold

    def wait_for_dark(self):
        while self.value():
            pass

    def wait_for_light(self):
        while not self.value():
            pass
